### PR TITLE
Delay load import features

### PR DESCRIPTION
**Pull request by tnrn9b**

### CODE REVIEW

**Comment by ryanmkurtz** [comment]
I haven't looked at the code yet, but my initial test for functionality looks good with the following simple program I wrote:

```
#include <Windows.h>

int main() {
	POINT p;
	wchar_t name[100];
	DWORD size = 100;

	CreateCursor(0, 1, 2, 3, 4, NULL, NULL); // user32.dll
	GetCursorPos(&p); // user32.dll
	GetUserName(name, &size); //advapi32.dll
}
```
The program was linked with the following flags to enable delay load importing:
`
/DELAYLOAD:"user32.dll" /DELAYLOAD:"advapi32.dll"`

Program can be downloaded here: [DelayLoadImport.zip](https://github.com/NationalSecurityAgency/ghidra/files/5776117/DelayLoadImport.zip)

Before:
before

After:
after

**Review by ryanmkurtz — CHANGES_REQUESTED** [review]

**Comment by ryanmkurtz** [comment]
I added a delayed import call into winsock which imports by ordinal, and got the following exception:

```
ERROR Error Importing File: Error importing file: DelayLoadImport.exe java.lang.NullPointerException
	at ghidra.app.util.opinion.PeLoader.processDelayImports(PeLoader.java:531)
	at ghidra.app.util.opinion.PeLoader.load(PeLoader.java:146)
	at ghidra.app.util.opinion.AbstractLibrarySupportLoader.doLoad(AbstractLibrarySupportLoader.java:347)
	at ghidra.app.util.opinion.AbstractLibrarySupportLoader.loadProgram(AbstractLibrarySupportLoader.java:83)
	at ghidra.app.util.opinion.AbstractProgramLoader.load(AbstractProgramLoader.java:112)
	at ghidra.plugin.importer.ImporterUtilities.importSingleFile(ImporterUtilities.java:400)
	at ghidra.plugin.importer.ImporterDialog.lambda$7(ImporterDialog.java:351)
	at ghidra.util.task.TaskBuilder$TaskBuilderTask.run(TaskBuilder.java:303)
	at ghidra.util.task.Task.monitoredRun(Task.java:124)
	at ghidra.util.task.TaskRunner.lambda$0(TaskRunner.java:93)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
  (ImporterUtilities.java:413) 
```

```
int main() {
	POINT p;
	wchar_t name[100];
	DWORD size = 100;

	CreateCursor(0, 1, 2, 3, 4, NULL, NULL); // user32.dll
	GetCursorPos(&p); // user32.dll
	GetUserName(name, &size); //advapi32.dll

	WSAGetLastError(); // Ws2_32.dll (import by ordinal)
}
```

[DelayLoadImportWinsock.zip](https://github.com/NationalSecurityAgency/ghidra/files/5776275/DelayLoadImportWinsock.zip)

File: Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/DelayImportDataDirectory.java [hunk 1]
```diff
@@ -169,6 +187,18 @@ public void markup(Program program, boolean isBinary, TaskMonitor monitor, Messa
 			addr = addr.add(descriptor.sizeof());
 		}
     }
+    
+    private void createExternalLibrary(Program program, String dllName) {
+    	try {
+		    program.getExternalManager().addExternalLibraryName(dllName, SourceType.IMPORTED);
+		} catch (DuplicateNameException | InvalidInputException e) {}
```
[the remark this answers is not in the document]
Comment (mumbel) [reply Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/DelayImportDataDirectory.java:194]: Any chance of you describing that in more detail, I didn't know this was an option

[the remark this answers is not in the document]
Comment (tnrn9b) [reply Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/DelayImportDataDirectory.java]: Import the eclipse code preferences for the Ghidra project and apply the formmater

**Comment by tnrn9b** [comment]
Reformatted and added structures for the delayed import by ordinal. Resolving the rest should be left to the end user or a script to be run post load.

**Comment by ryanmkurtz** [comment]
I did some work on this and it's in internal review.  I was able to simplify it quite a bit because to get ordinals to work, you just have to name the external references "Ordinal_n", and something that comes later in the load process will go out and replace that with the real function name.

**Pull request closed without merging**